cloudinit/sources/DataSourceScaleway.py

```diff
@@ -177,51 +186,152 @@ def __init__(self, sys_cfg, distro, paths):
         )
 
+    def _set_urls_on_ip_version(self, proto, urls):
+
+        if proto not in ["ipv4", "ipv6"]:
+            LOG.debug("Invalid IP version : %s", proto)
+            return []
+
+        filtered_urls = []
+        for url in urls:
+            # Numeric IPs
+            address = urlparse(url).netloc
+            if address[0] == "[":
+                address = address[1:-1]
+            addr_proto = socket.getaddrinfo(
+                address, None, proto=socket.IPPROTO_TCP
+            )[0][0]
+            if addr_proto == socket.AF_INET and proto == "ipv4":
+                filtered_urls += [url]
+                continue
+            elif addr_proto == socket.AF_INET6 and proto == "ipv6":
+                filtered_urls += [url]
+                continue
+
+        return filtered_urls
+
     def _get_data(self):
         if not on_scaleway():
             return False
 
         if self._fallback_interface is None:
             self._fallback_interface = net.find_fallback_nic()
-        try:
-            with EphemeralDHCPv4(
-                self._fallback_interface,
-                tmp_dir=self.distro.get_tmp_exec_path(),
-            ):
-                util.log_time(
-                    logfunc=LOG.debug,
-                    msg="Crawl of metadata service",
-                    func=self._crawl_metadata,
-                )
-        except (NoDHCPLeaseError) as e:
-            util.logexc(LOG, str(e))
-            return False
+        if self.has_ipv4:
+            try:
+                # DHCPv4 waits for timeout defined in /etc/dhcp/dhclient.conf
+                # before giving up. Lower it in config file and try it first as
+                # it will only reach timeout on VMs with only IPv6 addresses.
+                with EphemeralDHCPv4(
+                    self._fallback_interface,
+                    tmp_dir=self.distro.get_tmp_exec_path(),
+                ) as ipv4:
+                    ipv4_urls = self._set_urls_on_ip_version(
+                        "ipv4", self.metadata_urls
+                    )
+                    util.log_time(
+                        logfunc=LOG.debug,
+                        msg="Set api-metadata URL depending on "
+                        "IPv4 availability",
+                        func=self._set_metadata_url,
+                        args=(ipv4_urls,),
+                    )
+                    util.log_time(
+                        logfunc=LOG.debug,
+                        msg="Crawl of metadata service",
+                        func=self._crawl_metadata,
+                    )
+                    self.ephemeral_fixed_address = ipv4["fixed-address"]
+                    self.metadata["net_in_use"] = "ipv4"
+            except (
+                NoDHCPLeaseError,
+                ConnectionError,
+                ProcessExecutionError,
+            ) as e:
+                util.logexc(LOG, str(e))
+                self.has_ipv4 = False
+        if not self.has_ipv4:
```

Comment:
Using `self.has_ipv4 = False` to define control flow feels weird. Why not just return `true` at the end of the `try` block and allow flow to continue if an exception gets thrown?

Reply:
This has been put there to avoid a second DHCPv4 timeout when the `init` phase runs as it also calls `_get_data` and when executed with IPv6 only, we hit the 5 second dhcp timeout twice.
There maybe a better way to handle this though

Reply:
Are you saying that with this change that `has_ipv4` will be set during Local stage so that it doesn't have to timeout again in Network stage? If so I don't believe that's how this works, since this shouldn't run at Network, but maybe I'm missing something.

Reply:
I have sent you a copy of the `cloud-init.log` without the `if not self.has_ipv4` test where we clearly see the dhcp timeout appearing twice.
I believe that it is caused by the use of `EventType.BOOT` in `EventScope.NETWORK` that we use as we want to do the network configuration on each boot as the IP address may change b/w reboots.
Let me know if you need access to a test instance.

Reply:
Thanks! I'll take a look.

Reply:
Does ScaleWay use a dhcp server in the cloud environment? I wasn't expecting to see the exception:
```
cloudinit.net.dhcp.InvalidDHCPLeaseFileError: Cannot parse empty dhcp lease file /run/dhclient.lease
```

Is this is currently happening on all cloud-init instances on Scaleway (before this PR)? If so, that's a bug.

Since cloud-init is clearly not using the dhcp lease (it never got one), we could probably just bring up the interface with `EphemeralIPv4Network` to connect over the link local address, rather than using `EphemeralDHCPv4`. This will avoid the timeout issues we are seeing here.

Reply:
Let me be a bit clearer on the current and upcoming behavior.
Up until now, IPv4 was configured using DHCPv4, hence the use of EphemeralDHCPv4 in the previous version. The optional IPv6 address was statically defined. With the upcoming changes to our infrastructure, we can have 3 different situations : 

- DHCPv4 only
- DHCPv4 and SLAAC IPv6
- SLAAC IPv6 only

Since DHCPv4 is and will remain most of the cases we start by trying EphemeralDHCPv4 as previously.

Since we have no way of knowing beforehand which IP version will be available, we try EphemeralDHCPv4 even if it is expected to fail when SLAAC IPv6 only is available. In that case only, the EphemeralDHCPv4 will timeout and we will continue with EphemeralIPv6Network for IPv6 configuration.
 
The issue here is not knowing which of IPv4 and/or IPv6 is available before the start of the DataSource so we must try both cases, starting with EphemeralDHCPv4 which will work in the vast majority.

Reply:
Thanks for the context @karibou, this is much clearer now. Can we add this explanation to the comments please? This answers many of the questions I've had while reading this code, so it should help future readers as well.

Reply:
ok, will do

Reply:
Added comment to the code & commit log.